### DISCUSSION TITLE
DAI 2024-2025 - Java IOs benchmarking - Thomas Stäheli

### DISCUSSION THREAD

**Discussion by thomasstaheli** [heig-vd-dai-course/heig-vd-dai-course#398] [Show and tell]
# Java IOs

Here is the link of my repository : [Java ios repository (Thomas Stäheli)](https://github.com/thomasstaheli/heig-vd-dai-course-java-ios.git)

# Remarks

I'm the guy who had problems with the IntelliJ jetbrains IDE. I took your code from the solution to see at first if it was my code the problem.
But the problem is solved. I noticed that if I didn't load the maven files as suggested by the IDE, then my code wouldn't compile. So for all the other projects that I don't create myself, I have to click on the IDE's recommendation for the code to compile correctly.

# Observations 

### Binary Implemntation 

| Size | Write Time (ms) | Read Time (ms) | 
|-----------|-----------|-----------|
| 1 Byte     | 206    | 185 |
| 1 KiB       | 230    | 188 | 
| 1 MiB      | 1552  | 544| 
| 5 MiB      | 7319  | 1994  | 

### Buffered Binary Implementation 

| Size | Write Time (ms) | Read Time (ms) | 
|-----------|-----------|-----------|
| 1 Byte     | 184  |  193  |
| 1 KiB       | 185  |  194 | 
| 1 MiB      | 214  |  217 | 
| 5 MiB      | 309  |  306 | 

### Text Implementation

| Size | Write Time (ms) | Read Time (ms) | 
|-----------|-----------|-----------|
| 1 Byte     | 192 |  186 |
| 1 KiB       | 183 |  187 | 
| 1 MiB      | 283 |  499 | 
| 5 MiB      | 442 |  520 | 

### Buffered Text Implementation 

| Size | Write Time (ms) | Read Time (ms) | 
|-----------|-----------|-----------|
| 1 Byte     | 181 |  180 |
| 1 KiB       | 195 |  229 | 
| 1 MiB      | 223 |  224 | 
| 5 MiB      | 332 |  328 | 

# Answers to the questions

## Which type of stream is the most efficient for each use case?

As we can see in the tables, the better stream is the buffered one, for binary and text.

There is not a lot of difference when the size of the file is small, but when the file is bigger, there is a significant time difference.
We can see the the time difference in the Binary data section, with the 5 MiB file, there is a time difference of 7 secondes !

## Why is it more efficient than the other types of streams?

The buffered stream is better because he is doing less system calls. This method use more memory than the other one but save a lot of times.

## What is the difference between binary data and text data?

A binary file is composed of bytes and a text file is composed of characters.

## What is a character encoding?

This is the stage at which we convert the byte to characters, we can see this mapping in the ASCII table for example.

**Comment by ludelafo** [comment]
Thank you!

You can always find the solution in the solution repository if you ever need to compare your work with the official one here: https://github.com/heig-vd-dai-course/heig-vd-dai-course-solutions.

Do not hesitate to ask questions if something is not clear, either in this discussion or in the Q&A category of the GitHub Discussions here: https://github.com/orgs/heig-vd-dai-course/discussions/categories/q-a.

And thank you for the remarks. I was able to clone your project and to run it locally. I still had to change the SDK used for the project by going in the **Hamburger menu** (upper left icon next to the IntelliJ logo with the three lines) > Go in the **Files** section > **Project Structure** and then switch to a JDK available on my machine (_Temurin 21_ instead of the value _Temurin 21 (WSL)_ that is automatically set by your IDE).

It seems that we still aren't able to have a 100%, cross-platform, compatible way to share and develop project with people on different OSes but we are getting closer I hope..!

Thank you again, I'll see if this can be improved in the future.
